### PR TITLE
Extract semian ruby memory management from semian.c

### PR DESCRIPTION
**Pull request by dalehamel**
# What

Extract ruby GC related functions from semian.c

# Why

Part of omnibus refactor

### CODE REVIEW

**Comment by dalehamel** [comment]
Depends on https://github.com/Shopify/semian/pull/102

File: ext/semian/semian.c [hunk 1]
```diff
@@ -1,25 +1,4 @@
...
-#include <stdio.h>
-
-union semun {
-  int              val;    /* Value for SETVAL */
-  struct semid_ds *buf;    /* Buffer for IPC_STAT, IPC_SET */
-  unsigned short  *array;  /* Array for GETALL, SETALL */
-  struct seminfo  *__buf;  /* Buffer for IPC_INFO
-                             (Linux-specific) */
-};
+#include <semian.h>
```
Comment (dalehamel) [inline ext/semian/semian.c]: note should be pulled in by #102

File: ext/semian/semian_types.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+// see man semop
+union semun {
+  int              val;    /* Value for SETVAL */
+  struct semid_ds *buf;    /* Buffer for IPC_STAT, IPC_SET */
+  unsigned short  *array;  /* Array for GETALL, SETALL */
+  struct seminfo  *__buf;  /* Buffer for IPC_INFO
+                             (Linux-specific) */
+};
+
+// FIXME
```
Comment (dalehamel) [inline ext/semian/semian_types.h]: ignore this

**Comment by csfrancis** [comment]
Why don't you base this on #102 to make it easier to review?

**Comment by dalehamel** [comment]
@csfrancis done

File: ext/semian/semian_resource_alloc.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*
+For memory management operations of semian resources.
+*/
+#ifndef SEMIAN_RESOURCE_ALLOC_H
+#define SEMIAN_RESOURCE_ALLOC_H
+
+#include <semian.h>
```
Comment (csfrancis) [inline ext/semian/semian_resource_alloc.h]: You have a circular reference here. You can probably be more specific in what you need to include.

**Comment by csfrancis** [comment]
> I don't have much experience organizing C, but is this worth it? 

I have similar concerns. I understand your intend, but also don't like the idea of having to navigate through a bunch of files to figure out where something's located.

**Comment by dalehamel** [comment]
> I have similar concerns. I understand your intent, but also don't like the idea of having to navigate through a bunch of files to figure out where something's located.

These could possible be shared with semian_resource, but they don't have much to do with it.

I'd say of files here this is probably the set that is most suitable for merging (semian_resource and semian_resource_alloc).

I think that there's a clear separation between at least the following 3 concerns:

* the semian resource, that implements ruby operations
* semian ticket management
* direct interactions with sysV semaphores.

The function names also already lend themselves to this separation, as all semian_resource functions already include this in their name, likewise the ticket management stuff mentions tickets, and the semaphore stuff directly mentions semaphores.

By splitting things out into separate files, it also makes it easier to refactor complicated functions by pulling out the logic into statics, scoped only to that file.

If you don't agree, then we should probably close this entire set of PRs and revert the previous merge.

**Comment by dalehamel** [comment]
Decided offline that this belongs with the semian_resource code

**Pull request closed without merging**